drivers/lora/sx1276.c

```diff
@@ -0,0 +1,492 @@
+		.buffers = buf,
+		.count = 2,
+	};
+
+	if (!write) {
+		const struct spi_buf_set rx = {
+			.buffers = buf,
+			.count = 2
+		};
+
+		return spi_transceive(dev_data.spi, &dev_data.spi_cfg,
+				&tx, &rx);
+	}
+
+	return spi_write(dev_data.spi, &dev_data.spi_cfg, &tx);
+}
+
+int sx1276_read(u8_t reg_addr, u8_t *data, u8_t len)
+{
+	return sx1276_transceive(reg_addr, false, data, len);
+}
+
+int sx1276_write(u8_t reg_addr, u8_t byte)
+{
+	return sx1276_transceive(reg_addr | BIT(7), true, &byte, 1);
+}
+
+void SX1276WriteBuffer(u16_t addr, u8_t *buffer, u8_t size)
+{
+	int ret, i;
+
+	for (i = 0; i < size; i++) {
+		ret = sx1276_write(addr, buffer[i]);
+		if (ret < 0) {
+			LOG_ERR("Unable to read address: %x", addr);
+			return;
+		}
+	}
+}
+
+void SX1276ReadBuffer(u16_t addr, u8_t *buffer, u8_t size)
+{
+	int ret, i;
+	u8_t regval;
+
+	for (i = 0; i < size; i++) {
+		ret = sx1276_read(addr, &regval, 1);
+		if (ret < 0) {
+			LOG_ERR("Unable to read address: %x", addr);
+			return;
+		}
+
+		buffer[i] = regval;
+	}
+}
+
+void SX1276SetRfTxPower(int8_t power)
+{
+	int ret;
+	u8_t pa_config = 0;
+	u8_t pa_dac = 0;
+
+	ret = sx1276_read(SX1276_REG_PA_CONFIG, &pa_config, 1);
+	if (ret < 0) {
+		LOG_ERR("Unable to read PA config");
+		return;
+	}
+
+	ret = sx1276_read(SX1276_REG_PA_DAC, &pa_dac, 1);
+	if (ret < 0) {
+		LOG_ERR("Unable to read PA dac");
+		return;
+	}
+
+	pa_config = (pa_config & RF_PACONFIG_MAX_POWER_MASK) | 0x70;
+	pa_config &= RF_PACONFIG_PASELECT_MASK;
+
+#if defined CONFIG_PA_BOOST_PIN
+	pa_config |= RF_PACONFIG_PASELECT_PABOOST;
+
+	if (power > 17) {
+		pa_dac = (pa_dac & RF_PADAC_20DBM_MASK) | RF_PADAC_20DBM_ON;
+	} else {
+		pa_dac = (pa_dac & RF_PADAC_20DBM_MASK) | RF_PADAC_20DBM_OFF;
+	}
+
+	if ((pa_dac & RF_PADAC_20DBM_ON) == RF_PADAC_20DBM_ON) {
+		if (power < 5) {
+			power = 5;
+		} else if (power > 20) {
+			power = 20;
+		}
+
+		pa_config = (pa_config & RF_PACONFIG_OUTPUTPOWER_MASK) |
+			     ((power - 5) & 0x0F);
+	} else {
+		if (power < 2) {
+			power = 2;
+		} else if (power > 17) {
+			power = 17;
+		}
+
+		pa_config = (pa_config & RF_PACONFIG_OUTPUTPOWER_MASK) |
+			     ((power - 2) & 0x0F);
+	}
+#elif CONFIG_PA_RFO_PIN
+	if (power < -1) {
+		power = -1;
+	} else if (power > 14) {
+		power = 14;
+	}
+
+	pa_config = (pa_config & RF_PACONFIG_OUTPUTPOWER_MASK) |
+			     ((power + 1) & 0x0F);
+#endif
+	ret = sx1276_write(SX1276_REG_PA_CONFIG, pa_config);
+	if (ret < 0) {
+		LOG_ERR("Unable to write PA config");
+		return;
+	}
+
+	ret = sx1276_write(SX1276_REG_PA_DAC, pa_dac);
+	if (ret < 0) {
+		LOG_ERR("Unable to write PA dac");
+		return;
+	}
+}
+
+static int sx1276_lora_send(struct device *dev, u8_t *data, u32_t data_len)
+{
+	Radio.SetMaxPayloadLength(MODEM_LORA, data_len);
+
+	Radio.Send(data, data_len);
+
+	return 0;
+}
+
+static void sx1276_rx_done(u8_t *payload, u16_t size, int16_t rssi, int8_t snr)
+{
+	dev_data.rx_buf = payload;
+	dev_data.rx_len = size;
+
+	k_sem_give(&dev_data.data_sem);
+}
```

Comment:
After Rx, it is still in Rx mode.
Even if i increase the delay from receive to 10s, the Rx interrupt continues to occur every 1 second on the receive board.
``` diff
diff --git a/samples/drivers/lora/receive/src/main.c b/samples/drivers/lora/receive/src/main.c
index c68e17ad34..8c396e67d8 100644
--- a/samples/drivers/lora/receive/src/main.c
+++ b/samples/drivers/lora/receive/src/main.c
@@ -54,6 +54,6 @@ void main(void)
                LOG_INF("Received data: %s", log_strdup(data));
 
                /* Receive data at 1s interval */
-               k_sleep(1000);
+               k_sleep(10000);
        }
 }
```

My board has the following symptoms.
**1. Empty data**
```
***** Booting Zephyr OS version 2.0.99 *****
[00:00:00.221,000] <inf> sx1276: SX1276 Version:12 found
[00:00:01.375,000] <inf> lora_receive: Received data: helloworld
[00:00:02.377,000] <inf> lora_receive: Received data: helloworld
[00:00:03.379,000] <inf> lora_receive: Received data:
***** Booting Zephyr OS version 2.0.99 *****
[00:00:00.221,000] <inf> sx1276: SX1276 Version:12 found
[00:00:01.200,000] <inf> lora_receive: Received data: helloworld
[00:00:02.202,000] <inf> lora_receive: Received data:
***** Booting Zephyr OS version 2.0.99 *****
```

**2. SPI RXNE loop**
In this case, the board stops working.
```
(gdb) bt
#0  0x0800798a in LL_SPI_IsActiveFlag_RXNE (SPIx=0x40013000)
    at /work/zephyrproject_lora2/modules/hal/stm32/stm32cube/stm32l1xx/drivers/include/stm32l1xx_ll_spi.h:739
#1  ll_func_rx_is_not_empty (spi=0x40013000)
    at /work/zephyrproject_lora2/zephyr/drivers/spi/spi_ll_stm32.h:40
#2  spi_stm32_shift_m (data=0x20001e50 <spi_stm32_dev_data_1>, spi=<optimized out>)
    at /work/zephyrproject_lora2/zephyr/drivers/spi/spi_ll_stm32.c:124
#3  spi_stm32_shift_frames (data=0x20001e50 <spi_stm32_dev_data_1>, spi=<optimized out>)
    at /work/zephyrproject_lora2/zephyr/drivers/spi/spi_ll_stm32.c:187
#4  transceive (asynchronous=false, signal=0x0, rx_bufs=<optimized out>, 
    tx_bufs=0x20000bb0 <z_main_stack+584>, config=0x20000200 <dev_data+12>, 
    dev=<optimized out>)
    at /work/zephyrproject_lora2/zephyr/drivers/spi/spi_ll_stm32.c:439
#5  spi_stm32_transceive (dev=<optimized out>, config=0x20000200 <dev_data+12>, 
    tx_bufs=0x20000bb0 <z_main_stack+584>, rx_bufs=<optimized out>)
    at /work/zephyrproject_lora2/zephyr/drivers/spi/spi_ll_stm32.c:462
#6  0x0800565a in z_impl_spi_transceive (config=0x20000200 <dev_data+12>, 
    rx_bufs=0x20000bb8 <z_main_stack+592>, tx_bufs=0x20000bb0 <z_main_stack+584>, 
    dev=<optimized out>) at /work/zephyrproject_lora2/zephyr/include/drivers/spi.h:263
#7  spi_transceive (config=0x20000200 <dev_data+12>, 
--Type <RET> for more, q to quit, c to continue without paging--
    rx_bufs=0x20000bb8 <z_main_stack+592>, tx_bufs=0x20000bb0 <z_main_stack+584>, 
    dev=<optimized out>)
    at /work/zephyrproject_lora2/zephyr/samples/drivers/lora/receive/build/zephyr/include/generated/syscalls/spi.h:33
#8  sx1276_transceive (length=<optimized out>, data=<optimized out>, write=false, 
    reg=<optimized out>) at /work/zephyrproject_lora2/zephyr/drivers/lora/sx1276.c:212
#9  sx1276_read (reg_addr=reg_addr@entry=1 '\001', 
    data=data@entry=0x20000bdf <z_main_stack+631> " \005", len=len@entry=1 '\001')
    at /work/zephyrproject_lora2/zephyr/drivers/lora/sx1276.c:221
#10 0x08005684 in SX1276ReadBuffer (addr=<optimized out>, 
    buffer=0x20000bff <z_main_stack+663> "", size=<optimized out>)
    at /work/zephyrproject_lora2/zephyr/drivers/lora/sx1276.c:248
#11 0x0800951c in SX1276Read (addr=addr@entry=1)
    at /work/zephyrproject_lora2/modules/lib/loramac-node/radio/sx1276/sx1276.c:1267
#12 0x08009538 in SX1276SetOpMode (opMode=<optimized out>)
    at /work/zephyrproject_lora2/modules/lib/loramac-node/radio/sx1276/sx1276.c:1219
#13 0x08006094 in SX1276SetRx (timeout=timeout@entry=0)
    at /work/zephyrproject_lora2/modules/lib/loramac-node/radio/sx1276/sx1276.c:1035
#14 0x0800542a in sx1276_lora_recv (dev=<optimized out>, data=<optimized out>, timeout=-1)
    at /work/zephyrproject_lora2/zephyr/drivers/lora/sx1276.c:355
--Type <RET> for more, q to quit, c to continue without paging--
#15 0x080044e0 in lora_recv (timeout=-1, data=0x20000c40 <z_main_stack+728> "helloworld", 
    dev=0x20001fa0 <__device_sx1276_lora>)
    at /work/zephyrproject_lora2/zephyr/include/drivers/lora.h:133
#16 main () at /work/zephyrproject_lora2/zephyr/samples/drivers/lora/receive/src/main.c:48
(gdb) 
```

I consulted the [ping-pong](https://github.com/Lora-net/LoRaMac-node/blob/develop/src/apps/ping-pong/NucleoL152/main.c) source of `loramac-node` and found out that **`Sleep()`** is needed.
https://github.com/Lora-net/LoRaMac-node/blob/develop/src/apps/ping-pong/NucleoL152/main.c#L350-L364
``` c
void OnTxDone( void )
{
    Radio.Sleep( );
    State = TX;
}

void OnRxDone( uint8_t *payload, uint16_t size, int16_t rssi, int8_t snr )
{
    Radio.Sleep( );
    BufferSize = size;
    memcpy( Buffer, payload, BufferSize );
    RssiValue = rssi;
    SnrValue = snr;
    State = RX;
}
```


I think the source should be added as below.
``` diff
diff --git a/drivers/lora/sx1276.c b/drivers/lora/sx1276.c
index aedd0d841f..11bd93033c 100644
--- a/drivers/lora/sx1276.c
+++ b/drivers/lora/sx1276.c
@@ -327,6 +327,11 @@ void SX1276SetRfTxPower(int8_t power)
        }
 }
 
+static void sx1276_tx_done(void)
+{
+       Radio.Sleep();
+}
+
 static int sx1276_lora_send(struct device *dev, u8_t *data, u32_t data_len)
 {
        Radio.SetMaxPayloadLength(MODEM_LORA, data_len);
@@ -338,6 +343,8 @@ static int sx1276_lora_send(struct device *dev, u8_t *data, u32_t data_len)
 
 static void sx1276_rx_done(u8_t *payload, u16_t size, int16_t rssi, int8_t snr)
 {
+       Radio.Sleep();
+
        dev_data.rx_buf = payload;
        dev_data.rx_len = size;
 
@@ -472,6 +479,7 @@ static int sx1276_lora_init(struct device *dev)
 
        k_sem_init(&dev_data.data_sem, 0, UINT_MAX);
 
+       dev_data.sx1276_event.TxDone = sx1276_tx_done;
        dev_data.sx1276_event.RxDone = sx1276_rx_done;
        Radio.Init(&dev_data.sx1276_event);
```

console output
```
***** Booting Zephyr OS version 2.0.99 *****
[00:00:00.222,000] <inf> sx1276: SX1276 Version:12 found
[00:00:01.025,000] <inf> lora_receive: Received data: helloworld
[00:00:03.024,000] <inf> lora_receive: Received data: helloworld
[00:00:05.025,000] <inf> lora_receive: Received data: helloworld
...
...
[01:00:54.229,000] <inf> lora_receive: Received data: helloworld
[01:00:56.228,000] <inf> lora_receive: Received data: helloworld
```

[Symptoms with strange data](https://github.com/zephyrproject-rtos/zephyr/pull/18998#issuecomment-544810918) were also fixed by this fix.

Reply:
@KwonTae-young Great! Thanks for looking into this. I still haven't found time to look into this. Will test this change and integrate it very soon.

Thanks once again :)